🚀 2단계 - 수강신청(도메인 모델)

baeksangha opened the pull request:
## 2단계 - 수강 신청(도메인 모델)
### 수강 신청 기능 요구사항
* [x] 과정(Course)은 기수 단위로 운영하며, 여러 개의 강의(Session)를 가질 수 있다.
* [x] 강의는 시작일과 종료일을 가진다.
* [x] 강의는 강의 커버 이미지 정보를 가진다.
* [x] 이미지 크기는 1MB 이하여야 한다.
* [x] 이미지 타입은 gif, jpg(jpeg 포함),, png, svg만 허용한다.
* [x] 이미지의 width는 300픽셀, height는 200픽셀 이상이어야 하며, width와 height의 비율은 3:2여야 한다.
* [x] 강의는 무료 강의와 유료 강의로 나뉜다.
* [x] 무료 강의는 최대 수강 인원 제한이 없다.
* [x] 유료 강의는 강의 최대 수강 인원을 초과할 수 없다.
* [x] 유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.
* [x] 강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.
* [x] 강의 수강신청은 강의 상태가 모집중일 때만 가능하다.
* [x] 유료 강의의 경우 결제는 이미 완료한 것으로 가정하고 이후 과정을 구현한다.
* [x] 결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.
### 프로그래밍 요구사항
DB 테이블 설계 없이 도메인 모델부터 구현한다.
도메인 모델은 TDD로 구현한다.
단, Service 클래스는 단위 테스트가 없어도 된다.
다음 동영상을 참고해 DB 테이블보다 도메인 모델을 먼저 설계하고 구현한다.

## 코드 설명
2단계 요구사항을 보고 어디까지 구현하라는 것인지 도무지 감이 안잡혀서 ㅜ 코드 변경 사항이 너무 많아졌습니다. 코드 보시기 전에 description부터 참고 부탁드립니다 (_ _)
코드 작업을 요약하면 아래와 같습니다.
- 도메인 추가 (Session, Student)
- 어플리케이션 서비스, 도메인 서비스 추가
  - 어플리케이션 서비스: `SessionRegistrationService`, `SessionRegistrationValidator`
  - 도메인 서비스: `SessionService`, `StudentService`, `PaymentService`
  - 레포지토리: `SessionRepository`, `StudentRepository`, `PaymentRepository`
    - 구현체로 각각 `JdbcSessionRepository`, `JdbcStudentRepository`, `JdbcPaymentRepository`도 구현했습니다.

#### 코드 아키텍쳐
<img width="657" alt="image" src="https://github.com/user-attachments/assets/7cd5e383-f44c-47b3-ab7e-a7bbc9b7ebae" />

전체적인 코드 구현은 이렇고.. 내용은 ㅜ 하나하나 구현하려니 너무 길어져서 Session도 abstract class로 만들고 유료, 무료 강의 클래스 따로 구현하려다 그냥 Session 하나로 다 합쳤습니다. schema.sql 보시면 session 테이블에 컬럼으로 다 추가해 버렸어요.

하지만 코드를 다 갈아 엎을 각오가 되어 있으니 요구사항, 문법, 설계, 테스트 코드 등 전체적으로 고쳐야할 점 피드백 부탁드립니다💪
코드 양이 너무 길어 리뷰어님의 소중한 시간을 많이 잡아먹을 것 같아 미리 송구합니다.. 🥲🙇‍♂️

baeksangha committed:
feat: 기본 요구사항을 만족하는 도메인 모델 추가

baeksangha committed:
feat: 2단계 요구사항 구현

baeksangha committed:
feat: README 업데이트

baeksangha committed:
refactor: StudentService 패키지 이동

baeksangha committed:
feat: SessionRegistrationValidatorTest 코드 추가

baeksangha commented on src/test/java/nextstep/courses/domain/SessionRepositoryTest.java:
```diff
@@ -0,0 +1,60 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@JdbcTest
+@TestPropertySource(properties = {
+        "spring.sql.init.mode=always",
+        "spring.sql.init.schema-locations=classpath:schema.sql",
+        "spring.sql.init.data-locations="
+})
```
data.sql에 넣은 데이터 때문에 findById가 실패해서 추가해 주었습니다.

baeksangha commented on src/test/java/nextstep/courses/domain/SessionRepositoryTest.java:
```diff
@@ -0,0 +1,60 @@
+        Session session = new Session(LocalDate.of(2025, 4, 15), LocalDate.of(2025, 5, 15), new CoverImage(20 * 1024, "PNG", 600, 400), SessionStatus.READY);
+        sessionRepository.save(session);
+        Session savedSession;
+        try {
+            savedSession = sessionRepository.findById(1L);
+        } catch (EmptyResultDataAccessException e) {
+            savedSession = sessionRepository.findById(2L);
+        }
```
얘는 좀 답이 없는 구문인데요.. 괜찮으시다면 어떻게 하면 좋을지 조언 부탁드립니다.
조회만 돌리면 id: 1로 검색되고, 저장과 조회를 함께 돌리면 id: 2로 검색되어서 이렇게 해두었습니다.

neojjc2 replied:
진짜 고민을 많이 해주신것 같네요 😆 
단순하게 접근하면 `data.sql`에 초기 데이터 넣으시고 해주셔도 될 것 같습니다. 
이전 데이터가 남아있어서 그런데 기존 테스트 코드를 유지하신다면 `@BeforeEach`에 
이전 데이터가 남지 않도록 추가 해주시는 것도 간단한(?) 해결 방법이 될 수 있을 것 같습니다 🙇 
```java
@BeforeEach
void setUp() {
    sessionRepository = new JdbcSessionRepository(jdbcTemplate);
    try {
        jdbcTemplate.execute("DELETE FROM session");
        jdbcTemplate.execute("ALTER TABLE session ALTER COLUMN id RESTART WITH 1");
        System.out.println("테이블 초기화 성공");
    } catch (Exception e) {
        System.err.println("테이블 초기화 실패: " + e.getMessage());
        e.printStackTrace();
    }
}
```

baeksangha replied:
`ALTER TABLE session ALTER COLUMN id RESTART WITH 1`요 쿼리는 테스트할 때 유용하겠네요! 감사합니다. 말씀주신 방법대로 적용했습니다. https://github.com/next-step/java-lms/pull/723/commits/108a14725fd16726bf3a7f749364ce3defc201ad

neojjc2 commented:
#tag @neojjc2

neojjc2 commented on README.md:
```diff
@@ -13,4 +13,26 @@
+DB 테이블 설계 없이 도메인 모델부터 구현한다.
+도메인 모델은 TDD로 구현한다.
+단, Service 클래스는 단위 테스트가 없어도 된다.
+다음 동영상을 참고해 DB 테이블보다 도메인 모델을 먼저 설계하고 구현한다.
```
요구사항 정리도 잘 해주셨고 `Description`도 잘 작성해주셨습니다 💯

neojjc2 commented on src/main/java/nextstep/application/controller/SessionController.java:
```diff
@@ -0,0 +1,26 @@
+
+@RestController
+@RequestMapping("/students/{studentId}/sessions")
+public class SessionController {
```
어디까지 구현해야 할지 감이 안잡히실만도 하셨을 것 같습니다 😅 
`RestController`까지 생각하셨다면 충분히 그러셨을 것 같네요 
보통은 `Controller`까지는 아닌데 그래도 이왕 만들어주셨으니
더 짜임새 있는 테스트를 할 수 있을 것 같습니다 👍

neojjc2 commented on src/main/java/nextstep/application/service/SessionRegistrationService.java:
```diff
@@ -0,0 +1,35 @@
+        sessionService.addStudentAndSaveSession(session);
+        paymentService.payment(sessionId, studentId, session.getFee().longValue());
+        studentService.addSessionAndSaveStudent(student, sessionId);
+        return ResponseEntity.ok("수업 등록이 완료되었습니다.");
```
`Session Registration`의 역할을 서비스로 분리해주신 의도는 좋다고 생각합니다만
제가 보기엔 1단계 미션의 리팩토링 전의 `QnAService`를 보는 것 같았습니다 🤔 

이번 미션의 목적은 1단계 미션을 해보셔서 아시겠지만 도메인모델 객체가 핵심 비즈니스로직을
가져갈 수 있게하고, 그에 대한 테스트가 가능한 코드를 만드는게 목적입니다 😄 

`SessionRegistrationService`에 대한 테스트코드도 없고, 
지금 구조라면 어쩔수 없이 `mocking` 테스트를 작성해야 하는 구조네요

물론 **"Service 클래스는 단위 테스트가 없어도 된다."** 라는 요구사항이 있지만
이 의미는 `Service` 클래스는 단위테스트가 없어도 될 만큼 
도메인모델 테스트로 갈음할 수 있기 때문인 것으로 알고 있습니다. 

요 부분은 조금 개선 고민해주시면 좋을 것 같습니다 🙇
> 프로그래밍 요구사항 
> 단위 테스트 가능한 코드(핵심 비지니스 로직)를 도메인 모델 객체에 구현한다.

neojjc2 commented on src/main/java/nextstep/application/service/SessionRegistrationValidator.java:
```diff
@@ -0,0 +1,24 @@
+
+@Service
+public class SessionRegistrationValidator {
+    public void validate(Session session, Student student) {
```
`Session` 도메인 객체로 위임하면 좋을 것 같습니다 🙇

baeksangha replied:
감사합니다. 피드백 주신 부분 반영하여 코드 수정했습니다. https://github.com/next-step/java-lms/pull/723/commits/8d97cd0d99101f2186a3e037e7c9437fc60888e5

neojjc2 commented on src/main/java/nextstep/courses/domain/Course.java:
```diff
@@ -3,18 +3,11 @@
+    private final String title;
+    private final Long creatorId;
+    private final LocalDateTime createdAt;
+    private final LocalDateTime updatedAt;
```
`Course`가 `Session` 대한 관계정보를 가지고 있어야 할 것 같습니다 🙏 
> 기능 요구사항
> 과정(Course)은 기수 단위로 운영하며, 여러 개의 강의(Session)를 가질 수 있다.

neojjc2 commented on src/main/java/nextstep/courses/domain/CoverImage.java:
```diff
@@ -0,0 +1,69 @@
+
+import static nextstep.courses.domain.ImageType.SESSION_ALLOWED_IMAGE_TYPES;
+
+public class CoverImage {
```
`CoverImage` 구현 잘 해주셨습니다 💯

neojjc2 commented on src/main/java/nextstep/courses/domain/ImageType.java:
```diff
@@ -0,0 +1,9 @@
+    GIF, JPG, JPEG, PNG, SVG;
+
+    public static final List<ImageType> SESSION_ALLOWED_IMAGE_TYPES = List.of(GIF, JPG, JPEG, PNG, SVG);
+}
```
`SESSION_ALLOWED_IMAGE_TYPES`을 정의해서 관리하신 부분 좋습니다 👍 
소소한 의견이긴 합니다만 비즈니스 요건과 관련된 정의의므로 `CoverImage`가 가져가도 괜찮을 것 같네요 😄 
요 의견은 참고만 해주세요 🙏

baeksangha replied:
피드백 주셔서 감사합니다 🙇‍♂️ 요 건은 실무에서도 리뷰해주시는 분들마다 의견이 분분한 것 같네요.
CoverImage가 가져가도록 수정했습니다 ㅎㅎ https://github.com/next-step/java-lms/pull/723/commits/ac9869894e3a894ae06bc4b4b9ee9069d6aed320

neojjc2 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,71 @@
+
+    public void addStudent() {
+        studentsCount++;
+    }
```
`addStudent() `구조가 약간 어색합니다 🤔 

도메인모델 호출 관점에서 보면, 
`session.addStudent();` 만 호출하면 `count`가 늘어나는 구조인데, 어떤 제약도 없네요
위에서도 의견 드렸지만 `Session`에 수강을 할 때 적절한 `validation`이 `mandatory`하게 적용될 수 있도록
수정이 필요할 것 같습니다. 

데이터모델 관점에서 보면
지금 구조에서는 어떤 학생이 (`Student`) 이 강의를 (`Session`)을 수강했는지 알수가 없습니다 🤔 
의도하신 설계가 맞는지 궁금합니다 🙇 

그리고 데이터관리 관점에서도 이렇게 `Session`이 `count`를 관리하고 있다면 실시간 `count`관리가 어렵습니다.
신청, 취소등이 발생할 경우 계속 `count`를 수정하는 `update`가 발생하는데요
좋은 구조는 아니라고 생각합니다 🙇 

실제 비즈니스요건에서도 수강신청은 단시간에 트래픽이 집중적으로 몰리는 경우가 많습니다. 
최대한 경합을 줄여야 할 것 같은데 `count`에대한 접근이 병목이 될 수도 있을 것 같습니다. 

이 부분도 개선 고민 부탁 드리겠습니다 🙇

neojjc2 commented on src/main/java/nextstep/courses/infrastructure/JdbcSessionRepository.java:
```diff
@@ -0,0 +1,46 @@
+            );
+        }, id);
+    }
+}
```
3단계 미션까지 미리 해주셨네요 😅

baeksangha replied:
3단계 미션이 DB 적용인지 몰랐습니다. 🥲 다음 단계에서 뭘 할지도 미리 봐야겠군요.,

neojjc2 commented on src/main/java/nextstep/payments/domain/Payment.java:
```diff
@@ -26,4 +16,28 @@ public Payment(String id, Long sessionId, Long nsUserId, Long amount) {
+
+    public LocalDateTime getCreatedAt() {
+        return createdAt;
+    }
```
지불정보 (`Payment`)를 저장을 하고 있지만 `PaymentService`를 통하지 않고서는
알수가 없는 구조네요 🤔 
`Session`이든, `Student`이든 도메인모델 레벨에서 지불정보를 알고 있어야 할 것 같습니다. 

객체별로 `Repository`가 있고 `Service`를 통해서 객체를 핸들링 하는 구조인데
이 구조는 도메인모델 구조가 아닙니다. 
이 미션의 요구사항은 "DB 테이블 설계 없이 도메인 모델부터 구현한다." 입니다 😄 
근데 지금 상하님이 구현하신 내용은 DB에 결과가 저장되고 있다는 전제로 작업하신 것 같네요. 

사실 이번 미션에서는 `Repository`나 `Service`가 필요 없는 미션입니다. 
최대한 도메인모델을 정의하고 해당 도메인모델이 요구사항을 담당하고
그에 대한 테스트가 잘 되어있는지가 요지 입니다. 

`SessionService`정도는 엔트리로 있어도 무방 할 것 같긴 합니다만 😄 
최대한 도메인모델만 가지고 요구사항을 구현하는게 핵심이라고 봐주시면 될 것 같아요 🙇 

왜냐하면 이 정의하신 도메인모델 기준으로 `Repository`를 만드는게 3단계 미션이라서
지금은 도메인 모델을 정의 하고 해당 모델이 요구사항을 어떻게 담당할 수 있을지, 
필요하다면 어떤 객체가 더 필요하고, 원시객체 포장등을 통해 어떻게 더 세분화 해서 위임할지
관점에서 고민해주시면 될 것 같습니다 🙏

neojjc2 commented:
> ## 2단계 - 수강 신청(도메인 모델)
> ### 수강 신청 기능 요구사항
> * [x]  과정(Course)은 기수 단위로 운영하며, 여러 개의 강의(Session)를 가질 수 있다.
> * [x]  강의는 시작일과 종료일을 가진다.
> * [x]  강의는 강의 커버 이미지 정보를 가진다.
> * [x]  이미지 크기는 1MB 이하여야 한다.
> * [x]  이미지 타입은 gif, jpg(jpeg 포함),, png, svg만 허용한다.
> * [x]  이미지의 width는 300픽셀, height는 200픽셀 이상이어야 하며, width와 height의 비율은 3:2여야 한다.
> * [x]  강의는 무료 강의와 유료 강의로 나뉜다.
> * [x]  무료 강의는 최대 수강 인원 제한이 없다.
> * [x]  유료 강의는 강의 최대 수강 인원을 초과할 수 없다.
> * [x]  유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.
> * [x]  강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.
> * [x]  강의 수강신청은 강의 상태가 모집중일 때만 가능하다.
> * [x]  유료 강의의 경우 결제는 이미 완료한 것으로 가정하고 이후 과정을 구현한다.
> * [x]  결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.
> 
> ### 프로그래밍 요구사항
> DB 테이블 설계 없이 도메인 모델부터 구현한다. 도메인 모델은 TDD로 구현한다. 단, Service 클래스는 단위 테스트가 없어도 된다. 다음 동영상을 참고해 DB 테이블보다 도메인 모델을 먼저 설계하고 구현한다.
> 
> ## 코드 설명
> 2단계 요구사항을 보고 어디까지 구현하라는 것인지 도무지 감이 안잡혀서 ㅜ 코드 변경 사항이 너무 많아졌습니다. 코드 보시기 전에 description부터 참고 부탁드립니다 (_ _) 코드 작업을 요약하면 아래와 같습니다.
> 
> * 도메인 추가 (Session, Student)
> * 어플리케이션 서비스, 도메인 서비스 추가
>   
>   * 어플리케이션 서비스: `SessionRegistrationService`, `SessionRegistrationValidator`
>   * 도메인 서비스: `SessionService`, `StudentService`, `PaymentService`
>   * 레포지토리: `SessionRepository`, `StudentRepository`, `PaymentRepository`
>     
>     * 구현체로 각각 `JdbcSessionRepository`, `JdbcStudentRepository`, `JdbcPaymentRepository`도 구현했습니다.
> 
> #### 코드 아키텍쳐
> <img alt="image" width="657" src="https://private-user-images.githubusercontent.com/35217274/433388959-7cd5e383-f44c-47b3-ab7e-a7bbc9b7ebae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NzIwNDIsIm5iZiI6MTc0NDY3MTc0MiwicGF0aCI6Ii8zNTIxNzI3NC80MzMzODg5NTktN2NkNWUzODMtZjQ0Yy00N2IzLWFiN2UtYTdiYmM5YjdlYmFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDIzMDIyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViODY5Y2ViODFhZTBiODZjMjA5NzI0ZGM1NTU5ZGQ4NjZjNDA3YzA3YWNmYTI2YzUzMzI0M2QyMTkyNTM5NDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SImZCyDWXggC7Wg1fvfk8JWdYJ3GUM4Ct6rto4ksGto">
> 전체적인 코드 구현은 이렇고.. 내용은 ㅜ 하나하나 구현하려니 너무 길어져서 Session도 abstract class로 만들고 유료, 무료 강의 클래스 따로 구현하려다 그냥 Session 하나로 다 합쳤습니다. schema.sql 보시면 session 테이블에 컬럼으로 다 추가해 버렸어요.
> 
> 하지만 코드를 다 갈아 엎을 각오가 되어 있으니 요구사항, 문법, 설계, 테스트 코드 등 전체적으로 고쳐야할 점 피드백 부탁드립니다💪 코드 양이 너무 길어 리뷰어님의 소중한 시간을 많이 잡아먹을 것 같아 미리 송구합니다.. 🥲🙇‍♂️

@baeksangha 상하님 별말씀을 다 하십니다 😄 
이렇게 Description을 잘 작성해주셔서, 코드파악하는데 한결 수월했습니다.
리뷰 코멘트에도 남겼지만, 2, 3단계 미션을 같이 병행해주신 것 같아요 😆 
`Repository`, `Service`보다는 도메인모델을 잘 정의하고 해당 도메인이 비즈니스로직을 수행하고 역할을 가져갈 수 있게 하는게
2단계 미션의 주 내용이라고 이해해주시면 될 것 같습니다 🙇 

도메인서비스와 어플리케이션 서비스에 있는 비즈니스 로직들이  도메인 모델로 가면 될 것 같고, `Service`를 통한 접근 없이
도메인모델에서 필요한 상태들을 알 수 있는 구조가 되면 될 것 같습니다.

neojjc2 reviewed (changes requested):
안녕하세요 상하님 🙇 

2단계 미션 잘 진행해주셨습니다.
너무 어렵게 생각하신 것 같네요
고민의 깊이가 느껴집니다 😄 

의견 몇 가지 드렸는데요, 한번 보시고
개선 검토 해주시면 감사하겠습니다.

그럼 재 리뷰요청 기다리고 있겠습니다 🙇

baeksangha committed:
refactor: SessionRepositoryTest 테스트코드에 초기화 로직 추가

baeksangha committed:
refactor: 강의에 등록할 수 있는지 여부를 판단하는 로직을 session 객체에 위임

baeksangha committed:
refactor: CoverImage에서 사용할 수 있는 ImageType을 CoverImage 클래스 내에 저장

baeksangha committed:
refactor:
- 1급 컬렉션 객체 생성 (Sessions, Students)
- 수강 신청 로직을 session 객체 내에서 하도록 registerStudent 메서드 추가
- Course 도메인에 sessions 추가
- Session 도메인에 Students 추가, Student 도메인에 Sessions 추가

baeksangha commented:
@neojjc2 안녕하세요 준철님. 요구사항에 대한 구현도 엉망이고 2,3단계가 뒤섞인 와중에 섬세하게 리뷰해주셔서 너무 감사드립니다 🙇‍♂️ 피드백 보고 코드 수정하였습니다.
### 코드 수정 부분
#### SessionRegistrationService 로직 간소화
SessionRegistrationService에서 여러 서비스를 호출하던 부분을, `session.registerStudent`에서 한 번에 처리하도록 수정했습니다.
 `session.registerStudent`의 반환 타입은 Payment로 지정해 주었습니다.
#### 도메인 서비스라고 이름붙였던 클래스에 있던 로직을 도메인에 위임
addStudent, addSession 등 서비스 내에서 호출하던 코드들을 도메인 객체 내부로 위임하였습니다.
#### 그 외
- 일급 콜렉션 객체 생성 (Sessions, Students)
- Course 도메인에 sessions 추가

### 궁금한 사항, 고민되는 사항
- 아무래도 실무에서는 도메인 내부에는 책임이 없다시피 하고, 서비스 레이어에서 모든 로직을 처리하고 있기 때문에 평소에 작성하던대로 아예 분리해서 생각했던 것 같습니다. 준철님께서 남겨주신 리뷰를 보고 최대한 도메인에 책임을 전가하려고 하긴 했는데 피드백 주실 부분은 가감없이, 편하게 말씀 부탁드립니다.
- 저는 Session 클래스에도 해당 강의를 수강하는 학생들의 정보를 담은 Students 필드가 있어야 하고, Student 클래스에도 해당 학생이 수강하는 강의들의 정보가 담긴 Sessions 필드가 있어야 한다고 생각했습니다. (다대다 관계) 그러다 보니 Session, Student 클래스에 각각 학생과 강의 정보를 모두 넣어야주어야 한다고 생각했는데요, Session 도메인 내에서 registerStudent 메서드를 만들었다 보니 이 `registerStudent`메서드가 방대해진 것 같습니다. 작성할 때는 괜찮다고 생각했는데 여기서도 제가 실무에서 보던 익숙한 서비스 레이어의 코드 느낌이 나서.. 책임 분리를 위해 여기에서 더 할 수 있는 방안이 있을지 궁금합니다.

neojjc2 commented on src/test/java/nextstep/courses/domain/SessionRepositoryTest.java:
```diff
@@ -28,6 +27,13 @@ class SessionRepositoryTest {
         sessionRepository = new JdbcSessionRepository(jdbcTemplate);
+        try {
+            jdbcTemplate.execute("DELETE FROM session");
+            jdbcTemplate.execute("ALTER TABLE session ALTER COLUMN id RESTART WITH 1");
```
도움이 되셨다니 다행이네요 😄

neojjc2 commented on src/main/java/nextstep/application/service/SessionRegistrationService.java:
```diff
@@ -2,32 +2,27 @@
-        studentService.addSessionAndSaveStudent(student, sessionId);
-        return ResponseEntity.ok("수업 등록이 완료되었습니다.");
+        session.registerStudent(student);
+        return ResponseEntity.ok("수강 신청이 완료되었습니다.");
```
네 서비스에서 `register`는 딱 이정도가 좋은 것 같습니다
해당 비즈니스를 위해 필요한 값 들에 대한 사전 검증, 도메인 모델 간의 흐름제어, 결과 값 생성 정도 ?? 👍

neojjc2 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -26,25 +31,35 @@ public Session(Long id, LocalDate startDate, LocalDate endDate, CoverImage cover
     }
 
-    public void validateEnrollment(Long studentBudget) {
+    public void validateEnrollment(Student student) {
```
`Session`을 상속 받아 `PaidSession`, `FreeSession` 으로 
관심사를 분리해보는 것도 대안이 될 수 있을 것 같습니다 😄 
그럼 `Session` 입장에서는 추가 유형의 `Session`이 신설되도 수정이 없을 것 같고, `Paid`, `Free`도 
각자 검증에 대한 로직이 변경되어도 서로 영향이 없고, 수정범위도 줄어들 수 있을 것 같네요 😄

neojjc2 commented on src/main/java/nextstep/courses/domain/Sessions.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Sessions {
```
많이 개선 고민해주셨네요 😄

neojjc2 commented on src/main/java/nextstep/courses/domain/Students.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Students {
```
`Student`가 `Session`을 수강한 `NsUser`를 의미하는 도메인객체가 맞으신거죠?? 😄 
처음엔 학생객체여서 `NsUser`객체와 뭐가 다른거지?? 하고 리뷰했었는데,
점점 용도를 보니 `Session`을 수강한 내용의 역할을 하는 도메인모델로 보이네요. 

M:N관계는 1:N으로 푸는게 사실 데이터모델링의 교과서 적인 방법이긴 합니다 😄 

```markdown
+------------------+     +------------------+     +------------------+
|     Session      |     |   Registration   |     |     Student      |
+------------------+     +------------------+     +------------------+
| id               |1   *| id               |*   1| id               |
| startDate        |<--->| session_id       |<--->| name             |
| endDate          |     | student_id       |     | email            |
| coverImage       |     | register_date    |     | budget           |
| status           |     | status           |     |                  |
| type             |     | payment_info     |     |                  |
| capacity         |     |                  |     |                  |
| fee              |     |                  |     |                  |
+------------------+     +------------------+     +------------------+
```
텍스트로 표현하려니 약간 어색합니다만 위에 `Registration`이냐 `Enroll`같은 도메인객체가 하나 더 추가되서
둘의 관계를 담당하면 조금 더 정리된 모습일 것 같습니다 🙇

neojjc2 commented:
> @neojjc2 안녕하세요 준철님. 요구사항에 대한 구현도 엉망이고 2,3단계가 뒤섞인 와중에 섬세하게 리뷰해주셔서 너무 감사드립니다 🙇‍♂️ 피드백 보고 코드 수정하였습니다.
> 
> ### 코드 수정 부분
> #### SessionRegistrationService 로직 간소화
> SessionRegistrationService에서 여러 서비스를 호출하던 부분을, `session.registerStudent`에서 한 번에 처리하도록 수정했습니다. `session.registerStudent`의 반환 타입은 Payment로 지정해 주었습니다.
> 
> #### 도메인 서비스라고 이름붙였던 클래스에 있던 로직을 도메인에 위임
> addStudent, addSession 등 서비스 내에서 호출하던 코드들을 도메인 객체 내부로 위임하였습니다.
> 
> #### 그 외
> * 일급 콜렉션 객체 생성 (Sessions, Students)
> * Course 도메인에 sessions 추가
> 
> ### 궁금한 사항, 고민되는 사항
> * 아무래도 실무에서는 도메인 내부에는 책임이 없다시피 하고, 서비스 레이어에서 모든 로직을 처리하고 있기 때문에 평소에 작성하던대로 아예 분리해서 생각했던 것 같습니다. 준철님께서 남겨주신 리뷰를 보고 최대한 도메인에 책임을 전가하려고 하긴 했는데 피드백 주실 부분은 가감없이, 편하게 말씀 부탁드립니다.
> * 저는 Session 클래스에도 해당 강의를 수강하는 학생들의 정보를 담은 Students 필드가 있어야 하고, Student 클래스에도 해당 학생이 수강하는 강의들의 정보가 담긴 Sessions 필드가 있어야 한다고 생각했습니다. (다대다 관계) 그러다 보니 Session, Student 클래스에 각각 학생과 강의 정보를 모두 넣어야주어야 한다고 생각했는데요, Session 도메인 내에서 registerStudent 메서드를 만들었다 보니 이 `registerStudent`메서드가 방대해진 것 같습니다. 작성할 때는 괜찮다고 생각했는데 여기서도 제가 실무에서 보던 익숙한 서비스 레이어의 코드 느낌이 나서.. 책임 분리를 위해 여기에서 더 할 수 있는 방안이 있을지 궁금합니다.

@baeksangha 네 상하님 👍 그래도 많이 정리된 모습이네요 😄 

리뷰코멘트로도 남겼지만 M:N은 1:N으로 푸는게 사실 데이터모델링의 교과서적인 방식이긴 합니다. 
그리고 테스트코드 작성이 어렵거나, 스크립트 방식의 코드가 늘어난다면 
객체로 분리해서 관심사를 나눠야 하는 신호라고 생각합니다. 

사실 Service에서 객체들을 모아놓고 값들을 가져와서 하는 구현방식이 legacy코드 대부분이 그럴 것 같고
실제 요구사항을 구현하다보면 그게 빠르고 더 직관적일 때가 많은데요, 계속 그런방식으로 되다보니 나중엔
Service가 너무 방대해지고, Service에 있는 로직들이 여기저기 다른 Service에 중복되고
도메인객체는 단순히 Entity객체역할 밖에 하지 않고 Service내에 도메인들간에 의존도가 복잡해져서
뭐 하나 수정하려고 해도 전체를 분석해야 하는 상황을 매일 만나고 있습니다 😄

처음 개발이 시작되었을 때 대부분 데이터베이스 중심 아키텍처를 사용했고 (그 당시엔 그게 최선이었습니다) 
그 구조를 분석해서 개발을 하다보니 지금까지도  많은 곳에서 위 아키텍처로 되어있는 시스템이 많죠 🥲 
> https://assu10.github.io/dev/2024/05/06/clean-basic/

적절하게 도메인 객체들에게 역할과 책임을 위임함으로써 이런 문제들을 예방할 수 있습니다.
도메인 객체가 단순한 데이터 홀더가 아니라 자신의 비즈니스 로직을 캡슐화하면,
Service는 오케스트레이션에만 집중할 수 있고 중복 코드도 줄어듭니다.

이제 어느덧 오래된 서비스들도 등장하게 됐고 이제와서 돌아보니 대안으로 등장한 것들이 
도메인 중심, TDD등이라고 생각합니다. 

이런 개선이 당장은 시간이 좀 더 들더라도, 장기적으로는 유지보수성과 확장성 측면에서 큰 이득을 가져올 것이고 
특히 팀원 모두가 코드를 더 쉽게 이해하고 수정할 수 있게 될 것이라고 생각합니다. 
그럼 장애상황에서 불려가는 일도 적어질 것 같구요 😆
(실제 저도 실무에서 많이 개선하고 있고, 장점을 많이 느끼고 있습니다) 

위 부분은 다소 개인적인 의견일 수 있으니 참고 정도만 해주시고 (막상 쓰고 나서 보니 너무 라떼는.. 같은 느낌이네요 😭) 
이번 과정은 그 연습을 해보시는 거라 생각해주시면 될 것 같습니다 🙇

고민에 도움되셨기를 바라봅니다.

neojjc2 reviewed (approved):
안녕하세요 상하님 🙇 

개선고민 많이 해주셨네요 😄 
조금 더 정리되면 좋을 것 같아서
의견 드렸는데요, 3단계 진행하시면서 자연스럽게
고민하실 부분일 것 같아서
따로 재 리뷰 요청은 드리지 않겠습니다.
다음 단계 진행하시면서 고민해주시면 될 것 같아요 😄 

수고 많으셨습니다.
이만 `merge` 하겠습니다 🙇
